### DISCUSSION TITLE
Target rate sensors not turning on

### DISCUSSION THREAD

**Discussion by yebaws** [BottlecapDave/HomeAssistant-OctopusEnergy#579] [General]
I think I might have done an up date of the integration yesterday, not sure. But today, all of my sensors have failed to turn on. Entities are all enabled and the integration has started fine. I tried a restart and set up a test sensor, but that failed too.
Would much appreciate any ideas of how to troubleshoot this please. Everything else in HA working as expected, so don't thing it's an issue there

**Comment by BottlecapDave** [comment]
Nothing has changed around the target rate sensors work for a little while. Is the last evaluated attribute updating every minute? Is the target rates last evaluated attribute updatingnat least once a day? Have you performed a restart after updating the integration?

**Reply by yebaws** [reply]
Thanks for getting back. Since my post, I decided to just remove the integration, re-install and create all of my sensors again.
I remember there being a glitch in HA when I did the update, so thought this may have caused the issue. Re-downloading didn't work, so I went for the full re-install.
The test sensor that I have made since doing this has turned on, so fingers crossed that has fixed it!
Would be stuffed without it working. Thank you for your work, I think this is probably the most useful integration I have.

**Reply by BottlecapDave** [reply]
Yeah I've had various reports over the years of sensors just stopping as of HA has forgot about them. I've never managed to track down the root cause.

**Comment by fernando155** [comment]
Well... It happened to me again this time. First one I through it was update and takes some time, but after a week it happened again. It shows Target times last evaluated
27 November 2023 at 12:00:12
Last evaluated
27 November 2023 at 12:06:12

I checked a few and basically all off them are the same, same times.

**Reply by BottlecapDave** [reply]
hmm this is def starting to feel more like a HA issue as it feels very sporadic (among users). The last evaluated should be updated on every update, which is set to 1 minute. Do you have any errors in your logs? It might be worth following the FAQ to turn on debugging mode, leave for a couple of hours, turn it off and post the results here.

**Reply by yebaws** [reply]
My charge automations also all failed last night as the sensors didn't turn on again.

**Reply by BottlecapDave** [reply]
It might also be worth running the following (updating the sensor where appropriate) and seeing what comes back

```
{{ states.binary_sensor.octopus_energy_target_xxx.last_updated }}
```
and

```
{{ states.binary_sensor.octopus_energy_target_xxx.last_changed }}
```

**Reply by fernando155** [reply]
Sorry, what's that for? I'm happy to help, but I'm really inexpert sometimes.

I left debugging on. But as said, is been working fine for 4 entire days until today. So who knows if it will fail now or in 3 days time...

**Reply by BottlecapDave** [reply]
Sorry, they're templates which you can run in `developer tools` -> `template`. They will output HA internal timestamps. I'm wondering if they align with the `last evaluated` timestamps.

As for logs, in your case I would need debug logs for between 27/11/23 11:00-13:00 so I can see what HA was reporting just before it stopped updating.

**Reply by fernando155** [reply]
Except I restarted ha and I only can get logs after the restart, so everything is lost (didn't know that until now sorry). If it happens again I will post them

**Reply by yebaws** [reply]
Mine failed last night, again after a core update.

Entered this in developer tools template: 
```{{ states.binary_sensor.octopus_energy_target_eleven_thirty_to_seven_four_hours.last_updated }}```
Result was:
```2023-12-15 08:00:36.980897+00:00```

and 
```{{ states.binary_sensor.octopus_energy_target_eleven_thirty_to_seven_four_hours.last_changed
}}```

returns:
```2023-12-15 08:00:36.980897+00:00```

In integrations UI for Octopus Energy I get "Failed to set up" (presumably after core update, which was 2023.12.1 to 2023.12.3


```2023-12-15 08:00:36.980897+00:00```

Tarriff: Octopus Agile

**Reply by yebaws** [reply]
Again, up and running after a re-start. Have enabled debugging in the integration now, although I am sure I had enabled it last time and it seems to have turned off. So that I know next time, where do I find the debugging logs?

**Reply by BottlecapDave** [reply]
In the future if an integration fails, you can reload it manually without restarting. Unfortunately this can happen if the OE APIs can't be reached. HA stops debugging if you restart HA, but when active it can be found in the full HA logs in the normal location.

If the integration failed to setup, then this was probably why the sensor didn't turn on.

In a future update there will be a new attribute on target rate sensors that will indicate if rate information is incomplete and the result of target rate sensors not turning on.

**Reply by yebaws** [reply]
How would I re-load it manually please? Seems almost like the problem might be caused by the sequence in which things in HA re-start after a core update.

**Reply by BottlecapDave** [reply]
I personally haven't had any issues after updates (as it's like any other update), but because it needs access to the APIs, it's a bit of a luck of a draw if they're available.

To reload you go into the integration, find the entry that states it's failed to load and click the three dots next to it. This should open a menu, where you just click reload.

**Reply by yebaws** [reply]
Thanks, I'll try that next time

**Comment by yebaws** [comment]
Mine are working again after another restart. Will post logs here if / when they go down again

**Reply by alfwro13** [reply]
I have the same issue, my target rate binary sensor was off all night despite the fact it should have been On.
Was fixed by integration reload.
Im on v9.1.0 and HA 2023.12.1

**Reply by BottlecapDave** [reply]
To give me a chance of trying to track down this issue I need

* Logs from HA
* the last updated timestamp of the sensor. This can be retrieved by putting the following (updating for your sensor) in the template section of the developer tools
`{{ states.binary_sensor.octopus_energy_target_dishwasher.last_updated }}`
* The value of the target rates attribute on the sensor
* The value of `target_times_last_evaluated`
* The full config of the target rate sensors.
* your tariff

My assumption is the sensor doesn't turn on because there is a delay in retrieving rate information, and once present the sensor picks a best time which is in the past.

**Reply by alfwro13** [reply]
I will try to provide that info if I can replicate that issue in a control way - at night I just want it get working so I restart HA.
However as a workaround if anyone is having this issue I use this template that does the same thing, and use that in automation as a condition, automation is setup to run every 30 minutes. That template will render true if current time is the same as the date and time in the target_times start and false when end time is reached

```
{% set current_time = now().timestamp() %}
{% set target_times = state_attr('binary_sensor.octopus_energy_target_intermittent_best_charging_rates', 'target_times') %}

{% for target_time in target_times %}
  {% set start_time = as_timestamp(target_time['start']) %}
  {% set end_time = as_timestamp(target_time['end']) %}
  
  {% if current_time >= start_time and current_time < end_time %}
    true
  {% endif %}
{% endfor %}
false
```

**Comment by seanmaskey** [comment]
Hi
I'm new to HA; I have tried to follow instructions for setup & this query so apologies if ive missed something obvious. this Conversation seems relevant to the issue the has just developed for me.

I installed your very useful integration to run my heating and how water using octopus agile & updated to v9.1.
I started with a DHW immersion and have slowly added switches to control other appliances and most recently a Ground Source Heat pump. all worked well until a few days ago when the dependant Octopus energy rates display card failed to initialise consistently when restarting HA; it did so after another restart or I realised after  a few minutes with no action, but then (yesterday) stopped its self recovery.
The reason appears to be that the integration is not loading on restart & has to be done so manually.

Ive attached the log from recent restart  - on this occasion it required 2x manual reload.

does this give you enough info to make sense of the [Home assistant octopus energy .txt](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/files/13707475/Home.assistant.octopus.energy.txt)

Raspberry Pi 3b. 
Supervisor 
HA OS11.2
Core 2023 21.1
the integration diagnostics file is empty
HA Octopus Energy v9.1.0

Thanks

Sean
